### PR TITLE
added adaptive shrinking in lstm recipe

### PR DESCRIPTION
**Pull request by vijayaditya**
This PR is not ready for merge, it has been created to discuss the adaptive shrinkage schedule.

### CODE REVIEW

File: egs/wsj/s5/steps/nnet3/lstm/train.sh [hunk 1]
```diff
@@ -451,6 +453,12 @@ while [ $x -lt $num_iters ]; do
   ilr=$initial_effective_lrate; flr=$final_effective_lrate; np=$num_archives_processed; nt=$num_archives_to_process;
   this_learning_rate=$(perl -e "print (($x + 1 >= $num_iters ? $flr : $ilr*exp($np*log($flr/$ilr)/$nt))*$this_num_jobs);");
 
+  if [ "$adaptive_shrink" == "true" ]; then
+    this_shrink=$(perl -e "print $shrink + ($flr*$this_num_jobs/$this_learning_rate) * (1 - $shrink)")
```
Comment (danpovey) [inline egs/wsj/s5/steps/nnet3/lstm/train.sh]: I'm not convinced this is right.  We want (1-shrink) to vary proportional to this_learning_rate.   I initially intended that the specified shrinkage be the initial one.
So 
  this_shrink=$(perl -e "print 1.0 - (($this_learning_rate / ($num_jobs_initial * $ilr)) * (1 - $shrink));")

Comment (danpovey) [inline egs/wsj/s5/steps/nnet3/lstm/train.sh]: Actually, after thinking about this some more, I think it might be better to make it
this_shrink=$(perl -e "print 1.0 - (($this_learning_rate / $ilr)) * (1 - $shrink));")
This makes the shrink setting independent of the initial num-jobs, so that you could change the initial num-jobs without materially affecting the shrinkage rate.

Comment (vijayaditya) [inline egs/wsj/s5/steps/nnet3/lstm/train.sh]: this_learning_rate is a function of the num_initial_jobs, so to make this_shrink independent of num_initial_jobs we actually need

this_shrink=$(perl -e "print 1.0 - (($this_learning_rate / ($num_jobs_initial * $ilr)) * (1 - $shrink));")

Comment (danpovey) [inline egs/wsj/s5/steps/nnet3/lstm/train.sh]: The way you are doing it would make the initial shrinking rate independent of the num-jobs-initial, that's true, but it would make the final shrinking rate highly dependent on num-jobs-initial too.  So, for instance, between num-jobs-initial=1 and num-jobs-initial=2 there would be a factor of 2 difference in the final-probs.  I think it's more appropriate to do it the way I proposed, as in
 this_shrink=$(perl -e "print 1.0 - (($this_learning_rate / $ilr)) * (1 - $shrink));").
because this way, as we change the initial-num-jobs, the shrinking rate increases, which is actually appropriate because the initial actual learning rate is also increasing.  [this is in line with l2 regularization interpretation].

Comment (vijayaditya) [inline egs/wsj/s5/steps/nnet3/lstm/train.sh]: In addition to the rate of change, I am also worried about the initial shrink value. This, I believe, would affect learning to a greater extent. Further the shrink value is not going to reach 1 towards the end of the training in the above equation. We actually wanted $shrink to reach 1 as this would ensure we finally are optimizing the actual objective function.

So though we do not have an l2 regularization interpretation, how about keeping the original shrink value formula based on the effective_learning_rate rather than actual_learning_rate 
<code>
 this_shrink=$(perl -e "print $shrink + ($flr/($this_learning_rate/$this_num_jobs)) * (1 - $shrink)")
</code>

Comment (vijayaditya) [inline egs/wsj/s5/steps/nnet3/lstm/train.sh]: I take back my statement that initial shrink value is more important than rate.. I was looking at the results for various shrink values (See below plot) I think rate would be more important. So I implementing your formula..

shrink_vs_cvaccuracy

Comment (danpovey) [inline egs/wsj/s5/steps/nnet3/lstm/train.sh]: I think both my formula
 this_shrink=$(perl -e "print 1.0 - (($this_learning_rate / $ilr) * (1 - $shrink));")
which has an l2 regularization interpretation, and your formula
 this_shrink=$(perl -e "print $shrink + ($flr/($this_learning_rate/$this_num_jobs)) * (1 - $shrink)") 
which allows the shrinkage value to approach 1 (and might be suitable for larger data sets)
should be experimented with.  But I think your formula could be clarified by modifying the existing script to set a variable this_effective_lrate first, and then set a variable this_learning_rate afterwards; you can then use $this_effective_lrate in your formula instead of that ratio.

Comment (danpovey) [inline egs/wsj/s5/steps/nnet3/lstm/train.sh]: That graph could be interpreted as telling us that 0.95 shrinkage was too
high at the end, but 0.95, 0.98 and 0.99 were all in a reasonable range at
the beginning.  So it could be also be interpreted as saying that it's
important to aggressively decrease the shrinkage towards the end.  Anyway,
up to you.

Comment (vijayaditya) [inline egs/wsj/s5/steps/nnet3/lstm/train.sh]: OK, I will add both these till we have results on different databases. Looking at the above plot I think I will make your formula the default one.

**Comment by vijayaditya** [comment]
ready for merge.. current default option is no adaptive_shrinkage which reproduces the results with the existing lstm recipe.

File: egs/ami/s5/local/nnet3/run_lstm.sh [hunk 2]
```diff
@@ -102,6 +103,7 @@ if [ $stage -le 8 ]; then
     --cmvn-opts "--norm-means=false --norm-vars=false" \
     --initial-effective-lrate $initial_effective_lrate --final-effective-lrate $final_effective_lrate \
     --momentum $momentum \
+    --adaptive-shrink $adaptive_shrink \
```
Comment (danpovey) [inline egs/ami/s5/local/nnet3/run_lstm.sh]: You need to have "$adaptive_shrink" or this won't parse right.

**Comment by danpovey** [comment]
Could you please do some experiments before this is merged?  I don't want
to fill up the script with all kinds of options.  Let's just figure out
which schedule to use and then just make that the default.
Dan

**Comment by vijayaditya** [comment]
I am testing out the two types of shrinkage schedules. We will have some
results in 10 hours.

Vijay

File: egs/wsj/s5/steps/nnet3/lstm/train.sh [hunk 3]
```diff
@@ -449,9 +459,21 @@ while [ $x -lt $num_iters ]; do
...
 
   ilr=$initial_effective_lrate; flr=$final_effective_lrate; np=$num_archives_processed; nt=$num_archives_to_process;
-  this_learning_rate=$(perl -e "print (($x + 1 >= $num_iters ? $flr : $ilr*exp($np*log($flr/$ilr)/$nt))*$this_num_jobs);");
+  this_effective_learning_rate=$(perl -e "print ($x + 1 >= $num_iters ? $flr : $ilr*exp($np*log($flr/$ilr)/$nt));");
+  this_learning_rate=$(perl -e "print ($this_effective_learning_rate*$this_num_jobs);");
+  if [ z $adaptive_shrink ]; then
+    this_shrink=$shrink
+  else
+    if [ "$adaptive_shrink" == "type1" ]; then
+      this_shrink=$(perl -e "print 1.0 - (($this_learning_rate / $ilr) * (1 - $shrink));")
```
Comment (vijayaditya) [inline egs/wsj/s5/steps/nnet3/lstm/train.sh]: this_shrink=$(perl -e "print 1.0 - (($this_learning_rate / $ilr) * (1 - $shrink));")
which depends on the actual learning rate leads to zigzag and varying shrink values even for $shrink=0.95 and 0.98. Both the below training runs diverge

shrink_vs_cvaccuracy_iteration

Comment (vijayaditya) [inline egs/wsj/s5/steps/nnet3/lstm/train.sh]: red corresponds to $shrink=0.98 and blue 0.95

Comment (vijayaditya) [inline egs/wsj/s5/steps/nnet3/lstm/train.sh]: (rewriting the comment as it was hidden due to an updated pull request)

this_shrink=$(perl -e "print 1.0 - (($this_learning_rate / $ilr) * (1 - $shrink));")
which depends on the actual learning rate leads to zigzag and highly varying shrink values even for $shrink=0.95 and 0.98. Both the below training runs diverge

red corresponds to $shrink=0.98 and blue 0.95
Plot1 shows shrink values
and plot2 cv accuracies.


shrink_vs_cvaccuracy_iteration

File: egs/wsj/s5/steps/nnet3/lstm/train.sh [hunk 4]
```diff
@@ -449,9 +459,21 @@ while [ $x -lt $num_iters ]; do
...
 
   ilr=$initial_effective_lrate; flr=$final_effective_lrate; np=$num_archives_processed; nt=$num_archives_to_process;
-  this_learning_rate=$(perl -e "print (($x + 1 >= $num_iters ? $flr : $ilr*exp($np*log($flr/$ilr)/$nt))*$this_num_jobs);");
+  this_effective_learning_rate=$(perl -e "print ($x + 1 >= $num_iters ? $flr : $ilr*exp($np*log($flr/$ilr)/$nt));");
+  this_learning_rate=$(perl -e "print ($this_effective_learning_rate*$this_num_jobs);");
+  if [ -z "$adaptive_shrink" ]; then
+    this_shrink=$shrink
+  else
+    if [ "$adaptive_shrink" == "type1" ]; then
+      this_shrink=$(perl -e "print 1.0 - (($this_learning_rate / $ilr) * (1 - $shrink));")
```
Comment (vijayaditya) [inline egs/wsj/s5/steps/nnet3/lstm/train.sh]: (rewriting the comment as it was hidden due to an updated pull request)

this_shrink=$(perl -e "print 1.0 - (($this_learning_rate / $ilr) * (1 - $shrink));")
which depends on the actual learning rate leads to zigzag and highly varying shrink values even for $shrink=0.95 and 0.98. Both the below training runs diverge

red corresponds to $shrink=0.98 and blue 0.95
Plot1 shows shrink values
and plot2 cv accuracies.

shrink_vs_cvaccuracy_iteration

**Comment by vijayaditya** [comment]
Formula : <code> this_shrink=$(perl -e "print $shrink + ($flr/$this_effective_learning_rate) * (1 - $shrink)")</code>

Even with adaptive shrinkage the CV accuracies are no better than constant shrinkage of 0.99. The WERs are worse.

%WER 47.6   :  $shrink 0.95 + adaptive
%WER 47.1   :  $shrinkage 0.98 + adaptive
training diverges   :  $shrink 0.99 + adaptive
%WER 46.8 : $shrink 0.99 + constant 

(In the below plot 0.99 corresponds to the constant shrinkage run)
cv_shrink

**Comment by danpovey** [comment]
On Sun, Sep 27, 2015 at 6:16 AM, Vijayaditya Peddinti <
notifications@github.com> wrote:

> Formula : this_shrink=$(perl -e "print $shrink +
> ($flr/$this_effective_learning_rate) * (1 - $shrink)")
>
> Even with adaptive shrinkage the CV accuracies are no better than constant
> shrinkage of 0.99. The WERs are worse.
>
> %WER 47.6 : $shrink 0.95 + adaptive
> %WER 47.1 : $shrinkage 0.98 + adaptive
> training diverges : $shrink 0.99 + adaptive
>
Please implement that thing where you discard models from 'failed' jobs.  I
think the fact that shrink 0.99 + adaptive failed is likely just a fluke,
nothing systematic.  The values that I expect to work well are 0.99 and
0.995.  I think the 0.95 and 0.98 ones are just too strongly shrunk.
BTW, it would be nice to see (later) plots of the avg-deriv.  I'm hoping
that it will be more constant for the ones with adaptive shrinking.

Dan

**Comment by vijayaditya** [comment]
Ok, will do.

Vijay
